socketshark/receiver.py

```diff
@@ -79,29 +83,35 @@ async def ping_handler(self):
             if wait:
                 wait.cancel()
             if not self._stop:
-                self.shark.log.exception('unhandled exception in ping handler')
+                self.shark.log.debug('redis ping handler cancelled')
```

Comment:
Why the downgrade to `DEBUG` level?

Reply:
Because of the change to `ping_handler`, which executes a separate `_ping_handler` for each `redis_connection`, we now expect to receive the `CancelledError` exception. It is propagated from `ping_handler.cancel()`, so it isn't an unexpected error any more.

Reply:
I wasn't able to wrap my head around this yet. Can you walk me through what happens when you run it with `once` vs an unexpected scenario?

Is this comment still accurate?
```
except asyncio.CancelledError:  # Cancelled by stop()
```

If the readers run with `once` and finish, we call `ping_handler.cancel()`. Would that would raise `CancelledError` in each of the `_ping_handler`s? What's the state of `self._stop` at that point?

Reply:
When the `_reader` coroutine ends cleanly, the call to `ping_handler.cancel()` does propagate a `CancelledError` to each dependent `_ping_handler`. The state of `_stop` is still False. There are two relevant scenarios. 

1. `shark.run_service_receiver()` is called in `_run`, and an unexpected error is caught by `ServiceReceiver.reader`. 
  - The ping handler is stopped by the `CancelledError`
  - The `ServiceReceiver` is stopped by `SocketShark.shutdown()` invoked by `_run`. 
2. `shark.run_service_receiver(once=True)` is called, likely in tests.
  - `_reader` ends cleanly and the ping handler is stopped.
  - We may wish to invoke `run_service_receiver` again, so we should not set the `_stop` flag or close the connections.

As far as I can tell, any unknown exceptions are picked up in `reader`, where we will immediately abort, and run `ServiceReceiver.stop`.

Reply:
Just trying to understand this better:

> 1. shark.run_service_receiver() is called in _run, and an unexpected error is caught by ServiceReceiver.reader.
> - The ping handler is stopped by the CancelledError
> - The ServiceReceiver is stopped by SocketShark.shutdown() invoked by _run.

* If there is an unexpected error in  `reader` (https://github.com/closeio/socketshark/blob/23e58875e0cbebb263dab3f2790dbe849b1dbdaa/socketshark/receiver.py#L101), we log the exception but `ping_handler.cancel()` is not part of the exception handler. So we continue to `shutdown()` (via `_run`), which stops the service receiver, in which case the `_stop` flag will be true. What causes the ping handler to receive `CancelledError` in this scenario?
* If there is an unexpected error in `_reader` (https://github.com/closeio/socketshark/blob/23e58875e0cbebb263dab3f2790dbe849b1dbdaa/socketshark/receiver.py#L133) we keep looping unless it's called in tests using `once`, which seems fine.

> 2. shark.run_service_receiver(once=True) is called, likely in tests.
> - _reader ends cleanly and the ping handler is stopped.
> - We may wish to invoke run_service_receiver again, so we should not set the _stop flag or close the connections.

This makes sense, and I'm assuming this did not work well previously and would cause us to log an exception.

Reply:
> So we continue to shutdown() (via _run), which stops the service receiver, in which case the _stop flag will be true. What causes the ping handler to receive CancelledError in this scenario?

I see what you're saying now. I agree, the extra check for `self._stop` is unnecessary.

Removed in 4e3c6e8

Reply:
Or should we leave the check and cancel the ping handler explicitly in this scenario? If not, what would happen to the ping handler? Would it still be running?

Reply:
In this context, we're on the way to `shutdown`, so it would be killed when the process ends. The condition I removed just controls access to a low priority log level. 

What makes sense to me is, 2e99ab6, which ensures `ping_handler` is always cancelled at the end of `reader`. This keeps the changed behavior where `ping_handler` will only call `stop` if it encounters an error or a missed ping.